Implement token-based authentication

ohsayan opened the pull request:
This implements token-based authentication as described in #240. Closes #240.

ohsayan committed:
Add `Once` implementation

ohsayan committed:
Simplify `Once` implementation

ohsayan committed:
Add `AuthProvider` definition

ohsayan committed:
Implement `AuthProvider`

ohsayan committed:
Use smaller size for random bytes

A sample space of 40 provides enough randomness so that the
likelihood of two tokens being the same is almost entirely
ruled out. And even if they are, it shouldn't be a problem
since they correspond to different accounts, although the
possibility itself is as good as impossible due to the
possible number of permutations being 8.560685103E+94, so
we can safely ignore this "risk".

40 is a sensible balance between what clients have to send
and the level of security we're expecting to provide, than
64 bytes.

ohsayan committed:
Enable login and add login tests

ohsayan committed:
Fix try_from_slice impl

ohsayan committed:
Simplify action error propagation

ohsayan committed:
Fix length checks

Also improve panic messages in `dbtest` tests.

ohsayan committed:
Don't unnecessarily obtain arc clones

ohsayan committed:
Reduce duplication in `DdlError` handling

ohsayan committed:
Reduce duplication in strong action test impls

ohsayan committed:
Simplify flush routines using `StorageTarget`

ohsayan committed:
Fix `LocalSnapshot` and `RemoteSnapshot` implementations

ohsayan committed:
Fix tests

ohsayan committed:
Make flushing of keyspaces and tables generic over types

ohsayan committed:
Optimize table flush/unflush routines

ohsayan committed:
Make unflush routines generic over keyspaces and tables

ohsayan committed:
Fix unflush for volatile tables

ohsayan committed:
Add table restoration tests using bytemark

ohsayan committed:
Improve bytemark save/restore tests

ohsayan committed:
Version gate the storage engine

ohsayan committed:
Move system tables into separate namespace

ohsayan committed:
Add `ActionError` impls for `AuthError`

ohsayan committed:
Enable users to authenticate

ohsayan committed:
Implement `login`, `logout`, `claim` and `adduser` for `auth`

ohsayan committed:
Fix tests

ohsayan committed:
Enable auth to be configured

ohsayan committed:
Fix `auth login` and `auth claim` impls for unauth connections

ohsayan commented:
Can we have a bench here (without breaking hopefully)? @glydr bench

glydr commented:
I have added a benchmark request for this PR. Expect to hear back from me soon

ohsayan committed:
Fix length checks in auth subactions

ohsayan commented:
Looks like the `glydr bench` build failed. Try again @glydr bench

ohsayan commented:
The bench build failed. Can we retry? @glydr bench

glydr commented:
I have added a benchmark request for this PR. Expect to hear back from me soon

glydr commented:
The benchmark has completed. Review [the benchmark here](https://github.com/skytable/perf/blob/next/reports/result-09032022-125844.md)

ohsayan committed:
Simplify uninit array generation

The previous method used was a dirty hack due to limitations with
expressions in constant contexts, but the latest `rustc` lets
us greatly simplify this impl.

ohsayan committed:
Simplify dbtest macros

ohsayan committed:
Add separate `dbtest_func` macro for individual dbtest functions

ohsayan committed:
Enable using TLS on dbtest tests

ohsayan committed:
Enable authentication in dbtest tests

ohsayan committed:
Add authn tests

ohsayan committed:
Simplify tests

ohsayan committed:
Switch to Windows Server 2019

This might have been causing issues

Also, avoid path expansion on msys

ohsayan commented:
The failure on Windows is annoying me ... could it be because of actions/virtual-environments#4856?

ohsayan commented:
Windows builds fixed. It was msys all along

ohsayan committed:
Add authn tests

ohsayan committed:
Add `auth` syntax checks

ohsayan committed:
Add changelog and remove unused code

ohsayan commented:
CI is looking good. Ready to go! @glydr rmerge

glydr reviewed (approved):
Saw approval and instant merge request from @ohsayan
